content/architecture/storage-engine/prolly-tree.md

```diff
@@ -2,38 +2,234 @@
+
+#### At the Beginning
+
+When you insert a key at the beginning of the key space, the tree is modified along the left edge.
+
+![Prolly Tree Beginning Insert](../../.gitbook/assets/prolly-tree-front-insert.png)
+
+#### At the end
+
+When you insert a key at the end of the key space, the tree is modified along the right edge.
+
+![Prolly Tree End Insert](../../.gitbook/assets/prolly-tree-back-insert.png)
+
+#### In the middle
+
+When you insert a key in the middle of the key space, the tree is modified along a spline.
+
+![Prolly Tree Middle Insert](../../.gitbook/assets/prolly-tree-middle-insert.png)
+
+### Delete a key
+
+When you delete a key, the tree is modified under the same rules as an insert.
+
+![Prolly Tree Delete Key](../../.gitbook/assets/prolly-tree-delete.png)
+
+# Properties
+
+## History Independence
+
+A key property of a Prolly tree that enables fast diff and structural sharing is history independence. No matter which order you insert, update, or delete values, the Prolly tree is the same. This is best seen through example.
+
+Consider a map with 4 integer keys, `(1, 2, 3, 4)` pointing at the same values. Before hand we know the combination of keys `(1, 2)` will trigger a chunk boundary. We can know this beforehand because chunk boundaries are based on the size of the chunk and its contents. No matter which order we insert, update, or delete, if we will end up with a tree with two leaf nodes `(1, 2)`, and `(3,4)`. This also happens to be true if the keys have different values. The tree will be the same but the the chunks will have different addresses.
+
+Let's say we insert the chunks in sequential order.
+
+![Prolly Tree History Independence](../../.gitbook/assets/prolly-tree-history-independence-1.png)
+
+Then, let's say we insert the keys in reverse order.
+
+![Prolly Tree History Independence](../../.gitbook/assets/prolly-tree-history-independence-2.png)
+
+As you can see, we end up with the same Prolly Tree no matter which order we insert the values. It's a fun exercise to try and com e up with a sequence of inserts, updates, and deletes that result in a different tree that contains the same values. It's fun because you can't. The Prolly Tree algorithm always spits out the same tree.
```

Comment:
```suggestion
As you can see, we end up with the same Prolly Tree no matter which order we insert the values. It's a fun exercise to try and come up with a sequence of inserts, updates, and deletes that result in a different tree that contains the same values. It's fun because you can't. The Prolly Tree algorithm always spits out the same tree.
```